File: src/dbutils/session.py
# -*- mode: python; encoding: utf-8 -*-
#
# Copyright 2012 Jens Lindström, Opera Software ASA
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy of
# the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.

class Session(object):
    def __init__(self):
        self.__atexit = []
        self.storage = { "Repository": {},
                         "User": {},
                         "Commit": {},
                         "CommitUserTime": {},
                         "Timezones": {} }
        self.profiling = {}

        self.__user = None

    def atexit(self, fn):
        self.__atexit.append(fn)

    def close(self):
        for fn in self.__atexit:
            try: fn(self)
            except: pass
        self.__atexit = []

    def disableProfiling(self):
        self.profiling = None

    def recordProfiling(self, item, duration, rows=None, repetitions=1):
        if self.profiling is not None:
            count, accumulated_ms, maximum_ms, accumulated_rows, maximum_rows = self.profiling.get(item, (0, 0.0, 0.0, None, None))

            count += repetitions
            accumulated_ms += 1000 * duration
            maximum_ms = max(maximum_ms, 1000 * duration)

            if rows is not None:
                if accumulated_rows is None: accumulated_rows = 0
                if maximum_rows is None: maximum_rows = 0
                accumulated_rows += rows
                maximum_rows = max(maximum_rows, rows)

            self.profiling[item] = count, accumulated_ms, maximum_ms, accumulated_rows, maximum_rows

    @property
    def user(self):
        return self.__user

    def setUser(self, user):
        import auth
        assert not self.__user or self.__user.isAnonymous()
        self.__user = user

File: ui/src/resources/session.ts
/*
 * Copyright 2019 the Critic contributors, Opera Software ASA
 *
 * Licensed under the Apache License, Version 2.0 (the
); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an
 BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

import { immerable } from "immer"

import { primaryMap } from "../reducers/resource"

type SessionData = {
  user: null | number
  type: null | "normal" | "accesstoken"
  fields: FieldData[]
  providers: ProviderData[]
}

type SessionProps = {
  user: null | number
  type: null | "normal" | "accesstoken"
  fields: Field[]
  providers: Provider[]
}

class Session {
  [immerable] = true

  constructor(
    readonly user: null | number,
    readonly type: null | "normal" | "accesstoken",
    readonly fields: Field[],
    readonly providers: Provider[],
  ) {}

  static new(props: SessionProps) {
    return new Session(props.user, props.type, props.fields, props.providers)
  }

  static prepare(value: SessionData): SessionProps {
    return {
      ...value,
      fields: value.fields.map(Field.new),
      providers: value.providers.map(Provider.new),
    }
  }

  static reducer = primaryMap<Session, string>(
    "sessions",
    (_session) => "current",
  )

  get props(): SessionProps {
    return this
  }
}

type FieldData = {
  identifier: string
  label: string
  hidden: boolean
  description: string
}

type FieldProps = FieldData

export class Field {
  [immerable] = true

  constructor(
    readonly identifier: string,
    readonly label: string,
    readonly hidden: boolean,
    readonly description: string,
  ) {}

  static new(props: FieldProps) {
    return new Field(
      props.identifier,
      props.label,
      props.hidden,
      props.description,
    )
  }
}

type ProviderData = {
  identifier: string
  title: string
  account_id_label: string
}

type ProviderProps = ProviderData

export class Provider {
  [immerable] = true

  constructor(
    readonly identifier: string,
    readonly title: string,
    readonly accountIdLabel: string,
  ) {}

  static new(props: ProviderProps) {
    return new Provider(props.identifier, props.title, props.account_id_label)
  }
}

export default Session
